crates/services/relayer/src/service.rs

```diff
@@ -148,17 +149,20 @@ where
 impl<P, D> RunnableService for Task<P, D>
 where
     P: Middleware<Error = ProviderError> + 'static,
-    D: RelayerDb + 'static,
+    D: RelayerDb + Clone + 'static,
 {
     const NAME: &'static str = "Relayer";
 
-    type SharedData = SharedState;
+    type SharedData = SharedState<D>;
     type Task = Task<P, D>;
 
     fn shared_data(&self) -> Self::SharedData {
         let synced = self.synced.subscribe();
 
-        SharedState { synced }
+        SharedState {
+            synced,
+            database: self.database.clone(),
```

Comment:
I think it is better to put `Database` into `Arc`. It is used everywhere only to read except `insert_messages`(but you use db transaction inside, so it also could be read).

Reply:
It's up the the generic provider to provide a Clonable database and they all are clone